### PR TITLE
feat: Add partial support for `from_json`

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

`from_json` is a widely used Spark SQL function.

## What changes are included in this PR?

- Add very basic support for `from_json` for structs and a handful of primitive types.
- Add microbenchmarks

## What changes are NOT included in this PR?

- Support for arrays and maps
- Support for edge cases (numerics in single or double quotes, for example)
- Fuzz testing

## How are these changes tested?

New tests.

## Benchmark Results

```
OpenJDK 64-Bit Server VM 11.0.22+7-LTS on Mac OS X 14.6.1
Apple M3 Max
from_json - simple primitives:            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                5304           5356          73          0.2        5058.6       1.0X
SQL Parquet - Comet (Scan)                         5456           5484          40          0.2        5203.2       1.0X
SQL Parquet - Comet (Scan, Exec)                    310            322          13          3.4         295.3      17.1X

OpenJDK 64-Bit Server VM 11.0.22+7-LTS on Mac OS X 14.6.1
Apple M3 Max
from_json - all primitive types:          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                1716           1729          19          0.6        1636.3       1.0X
SQL Parquet - Comet (Scan)                         1713           1719           8          0.6        1633.9       1.0X
SQL Parquet - Comet (Scan, Exec)                    776            777           1          1.4         740.0       2.2X

OpenJDK 64-Bit Server VM 11.0.22+7-LTS on Mac OS X 14.6.1
Apple M3 Max
from_json - with nulls:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                3293           3315          31          0.3        3140.9       1.0X
SQL Parquet - Comet (Scan)                         3322           3362          56          0.3        3168.3       1.0X
SQL Parquet - Comet (Scan, Exec)                    264            297          35          4.0         252.0      12.5X

OpenJDK 64-Bit Server VM 11.0.22+7-LTS on Mac OS X 14.6.1
Apple M3 Max
from_json - nested struct:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                6901           6908          11          0.2        6580.9       1.0X
SQL Parquet - Comet (Scan)                         6873           6925          73          0.2        6554.5       1.0X
SQL Parquet - Comet (Scan, Exec)                    421            457          34          2.5         401.8      16.4X

OpenJDK 64-Bit Server VM 11.0.22+7-LTS on Mac OS X 14.6.1
Apple M3 Max
from_json - field access:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                5555           5599          61          0.2        5298.0       1.0X
SQL Parquet - Comet (Scan)                         5530           5559          41          0.2        5274.0       1.0X
SQL Parquet - Comet (Scan, Exec)                    255            292          36          4.1         242.9      21.8X
```

### CODE REVIEW

**Comment by coderfender** [comment]
This is awesome ! I believe we have an unwanted cargo.lock file in the PR @andygrove

**Comment by andygrove** [comment]
> This is awesome ! I believe we have an unwanted cargo.lock file in the PR @andygrove

The `cargo.lock` file has been updated due to the new dependency on `serde_json`.

File: native/core/src/execution/expressions/strings.rs [hunk 1]
```diff
@@ -98,3 +99,28 @@ impl ExpressionBuilder for RlikeBuilder {
         }
     }
 }
+
+/// Builder for FromJson expressions
```
Comment (mbutrovich) [inline native/core/src/execution/expressions/strings.rs]: Not a terribly useful comment IMO. Code is already self-documenting.

File: spark/src/main/scala/org/apache/comet/serde/structs.scala [hunk 2]
```diff
@@ -167,3 +167,67 @@ object CometStructsToJson extends CometExpressionSerde[StructsToJson] {
     }
   }
 }
+
+object CometJsonToStructs extends CometExpressionSerde[JsonToStructs] {
+
+  override def getSupportLevel(expr: JsonToStructs): SupportLevel = {
+    // this feature is partially implemented and not comprehensively tested yet
```
Comment (mbutrovich) [inline spark/src/main/scala/org/apache/comet/serde/structs.scala:174]: Do you have criteria for when this would change? You allude to arrays, maps, and fuzz tests in the PR description.

File: native/spark-expr/src/json_funcs/from_json.rs [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+    let DataType::Struct(fields) = schema else {
+        return Err(datafusion::common::DataFusionError::Execution(
+            "from_json requires struct schema".to_string(),
+        ));
+    };
+
+    // Build struct array by parsing each JSON string
+    let num_rows = string_array.len();
+
+    // Create builders for each field
```
Comment (mbutrovich) [inline native/spark-expr/src/json_funcs/from_json.rs]: "What" not "why" comment is not needed IMO.

File: native/spark-expr/src/json_funcs/from_json.rs [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+    })?;
+
+    // Schema must be struct
+    let DataType::Struct(fields) = schema else {
+        return Err(datafusion::common::DataFusionError::Execution(
+            "from_json requires struct schema".to_string(),
+        ));
+    };
+
+    // Build struct array by parsing each JSON string
```
Comment (mbutrovich) [inline native/spark-expr/src/json_funcs/from_json.rs]: Possible misplaced comment, or just generally refers to block of code below.

File: native/spark-expr/src/json_funcs/from_json.rs [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+                Err(_) => {
+                    // Parse error -> struct with null fields (PERMISSIVE mode)
+                    *struct_null = true;
+                    append_null_to_all_builders(&mut field_builders);
+                }
+            }
+        }
+    }
+
+    // Finish builders
```
Comment (mbutrovich) [inline native/spark-expr/src/json_funcs/from_json.rs]: "what" not "why"

File: native/spark-expr/src/json_funcs/from_json.rs [hunk 6]
```diff
@@ -0,0 +1,652 @@
...
+    let arrays: Vec<ArrayRef> = field_builders
+        .into_iter()
+        .map(finish_builder)
+        .collect::<Result<Vec<_>>>()?;
+
+    // Create null buffer from struct_nulls
```
Comment (mbutrovich) [inline native/spark-expr/src/json_funcs/from_json.rs]: "what" not "why"

File: spark/src/main/scala/org/apache/comet/serde/structs.scala (hunk 2 above)
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/serde/structs.scala:174]: This is mostly about testing. I mostly only have tests for the happy path so far i.e. correctly formatted JSON. Spark is very forgiving when parsing JSON.

**Review by mbutrovich — APPROVED** [review]
Thanks @andygrove! This will be a great expression to support.

**Comment by comphead** [comment]
Thanks @andygrove great start, it doesn't support nested structs?

**Comment by andygrove** [comment]
> Thanks @andygrove great start, it doesn't support nested structs?

It does support nested structs. See `test("from_json - nested struct")`.

**Review by comphead — COMMENTED** [review]
Can we also have a test with valid json and wrong schema?

**Review by comphead — APPROVED** [review]
Thanks @andygrove 💪

**Comment by andygrove** [comment]
Thanks for the reviews @mbutrovich and @comphead. I went ahead and merged despite the Spark 4 / hive-1 test failure which is happening on many PRs now. I filed https://github.com/apache/datafusion-comet/issues/2946 to get this resolved. The test is not related to `from_json` support.
